# Задание 2.1.1


Центрирование
 Обрезку
 Нормализацию
 Вращение 

# Задание 2.1.2:


### Распределите этапы решения задач машинного обучения в правильном порядке: 
"Определение требований;предобраобтка;моделирование;тестирование;внедрение"


# Задание 2.2.1:

### Какого типа признак «вес планшета»?
"Числовой верно"

# Задание 2.2.2

### Выберите все категориальные признаки:
Марка машины
Наличие тачпада у ноутбука

# Задание 2.2.3

### К какому типу данных относится оцифрованный снимок с пленочной камеры?
Графический (изображение) верно

# Задание 2.2.4

### Выберите, какие задачи можно решать с типом алгоритмов «Тексты»:
- [ ]  Искать ключевые слова
- [ ]  Переводить на другой язык

# Задание 2.2.5

### В неком чате Telegram есть масса сообщений. В результате задачи Х мы получили список наиболее интересных и важных тем из этих сообщений. Определите тип задачи Х, который применяется к тексту:
Поиск фактов
Суммаризация верно

# Задание 2.2.6

### К какому типу данных относится граф пользователей сети Facebook:
Смешанный тип верно

# Задание 2.2.7

### Выберите примеры графического типа данных (изображений):
Список позиций в чеке
Скан мозга, сделанный с помощью МРТ

# Задание 2.2.8

### Выберите примеры текстовых данных:
Письмо в почте
Пост в Facebook (без вложений)



# # **2.3. Практика**

In [3]:
import pandas as pd
data = pd.read_csv('data_flats.csv',sep=";")
data.head()

id  full_sq  life_sq  floor           sub_area  preschool_quota  \
0   1       43     27.0    4.0           Bibirevo           5001.0   
1   2       34     19.0    3.0  Nagatinskij Zaton           3119.0   
2   3       43     29.0    2.0     Tekstil'shhiki           1463.0   
3   4       89     50.0    9.0             Mitino           6839.0   
4   5       77     77.0    4.0          Basmannoe           3240.0   

   preschool_education_centers_raion  school_quota  \
0                                  5       11065.0   
1                                  5        6237.0   
2                                  4        5580.0   
3                                  9       17063.0   
4                                  7        7770.0   

   school_education_centers_raion  school_education_centers_top_20_raion  ...  \
0                               5                                      0  ...   
1                               8                                      0  ...   
2                               7                                      0  ...   
3                              10                                      0  ...   
4                               9                                      0  ...   

   shopping_centers_km  office_km  additional_education_km  preschool_km  \
0             0.648488   0.637189                 0.947962      0.177975   
1             0.519311   0.688796                 1.072315      0.273345   
2             1.486533   1.543049                 0.391957      0.158072   
3             0.599914   0.934273                 0.892674      0.236455   
4             0.429052   0.077901                 0.810801      0.376838   

   big_church_km  church_synagogue_km  theater_km  museum_km    ecology  \
0       0.625783             0.628187   14.053047   7.389498       good   
1       0.967821             0.471447    6.829889   0.709260  excellent   
2       3.178751             0.755946    4.273200   3.156423       poor   
3       1.031777             1.561505   16.990677  16.041521       good   
4       0.378756             0.121681    1.112486   1.800125  excellent   

   price_doc  
0    5850000  
1    6000000  
2    5700000  
3   13100000  
4   16331452  

[5 rows x 56 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30469 entries, 0 to 30468
Data columns (total 56 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     30469 non-null  int64  
 1   full_sq                                30469 non-null  int64  
 2   life_sq                                24086 non-null  float64
 3   floor                                  30302 non-null  float64
 4   sub_area                               30469 non-null  object 
 5   preschool_quota                        23781 non-null  float64
 6   preschool_education_centers_raion      30469 non-null  int64  
 7   school_quota                           23784 non-null  float64
 8   school_education_centers_raion         30469 non-null  int64  
 9   school_education_centers_top_20_raion  30469 non-null  int64  
 10  hospital_beds_raion                    16029 non-null  float64
 11  he

# Задание 2.4.1

### Какие бывают проблемы с данными?
Грязные данные
Неполные данные
Утечка данных

# Задание 2.4.2

### Какие данные мы называем неполными?

В данных есть пропуски; Нет данных за какой-то период; В данных смешиваются буквы из разных языков; Данные есть только для части задач

## Задание 2.4.3

### Выберите примеры грязных данных:


ans = В данных есть пропуски; Некоторые символы написаны не тем языком

## Задание 2.4.4

### Что такое утечка в данных?


ans = В данных есть признаки, от которых явно зависит целевая переменная

## Задание 2.4.5

### Вам прислали текст на русском языке, где попадаются китайские иероглифы. Часть параметров задаётся в метрах и сантиметрах, а некоторые — в ли, уин и чи. К какому типу данных можно отнести полученную информацию?


ans = Грязные данные


## Задание 2.4.6

## Что такое разметка данных?


ans = Процесс снижения влияния человеческого фактора на данные

## Задание 2.4.7

### Что из этого не является примером плохой разметки?


ans = Отсутствуют некоторые значения нецелевых переменных в данных



# 2.5. Работа с пропусками

## Задание 2.5.1

### Что такое данные с пропусками?


ans - Данные, в которых пропущена часть значений для нецелевой переменной верно


## Задание 2.5.2

### Какие из следующих случаев являются пропусками?

ans = Не все респонденты при опросе указали возраст; Часть значений указаны как nan; При передаче данных пришло 700 строк из 1000

## Задание 2.5.3

### Как можно обработать пропуски?

ans = Вернуть данные с пропусками; Удалить данные с пропусками; Заменить пропущенные данные константой; Обучить алгоритм, заменяющий пропуски


# 2.6. Практика

In [19]:
import pandas as pd
import numpy as np
test_data = pd.DataFrame([[1, 2, np.nan], [3, np.nan, 4],
                          [0, 1, 2]], columns=['one', 'two', 'three'])

In [12]:
test_data.isna()

one    two  three
0  False  False   True
1  False   True  False
2  False  False  False

In [13]:
test_data.dropna(axis=0)

one  two  three
2    0  1.0    2.0

In [14]:
test_data.fillna(-1)

one  two  three
0    1  2.0   -1.0
1    3 -1.0    4.0
2    0  1.0    2.0

## Задание 2.6.1

### С помощью какого метода в Pandas можно обнаружить наличие пропусков?


isna()

## Задание 2.6.2

### С помощью какого метода в Pandas можно удалить строки с пропусками?

xddropna()


## Задание 2.6.3

### С помощью какого метода в Pandas можно заполнить пропуски какими-либо значениями?

fillna()


## Задание 2.6.4

### Когда мы пишем в методе axis = 1, то подразумеваем:

применение операции к столбцам верно


## Задание 2.6.5

### Что происходит, когда применяем .fillna(np.median())?

заполняем пропуски значением, которое находится посередине (является медианой) для всех значений строки с пропуском верно


# Задание 2.7 Начальная обработка

## Задание 2.7.1
### Что такое нормализация данных?

ans = Процесс приведения данных к одному масштабу

## Задание 2.7.2
### Выберите методы нормализации

ans = MinMax; Стандартная

## Задание 2.7.3
### В каком случае лучше просто удалить данные, чем как-либо их обрабатывать?

ans = Признак не несёт полезной информации верно

# 2.8. Практика

In [24]:
test_data=test_data.fillna(0)


In [25]:
test_data

one  two  three
0    1  2.0    0.0
1    3  0.0    4.0
2    0  1.0    2.0

In [26]:
from sklearn.preprocessing import MinMaxScaler

In [27]:
scaler = MinMaxScaler()
scaler.fit_transform(test_data)

array([[0.33333333, 1.        , 0.        ],
       [1.        , 0.        , 1.        ],
       [0.        , 0.5       , 0.5       ]])

In [28]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit_transform(test_data)

array([[-0.26726124,  1.22474487, -1.22474487],
       [ 1.33630621, -1.22474487,  1.22474487],
       [-1.06904497,  0.        ,  0.        ]])

In [29]:
test_data.drop('one', axis=1)

two  three
0  2.0    0.0
1  0.0    4.0
2  1.0    2.0

# 2.9. Визуализация

## Задание 2.9.1
### С помощью какого графика нагляднее всего можно показать нормализованные данные?

ans = Гистограмма верно

## Задание 2.9.2
### Что позволяет сделать гистограмма?

ans = распределить данные; выделить неверные значения; сделать данные понятными визуально



## Задание 2.9.3
### Выберите из списка значения, которые имеют абсолютную корреляцию выше 0.5?

ans = 0.7; 0.69

## Задание 2.9.4
### Если зависимость большая и линейная, то какие будут значения корреляции?

ans = Большие